### PR TITLE
*Session attributes handling enhancement and TcpSocketSession cleanup

### PR DESCRIPTION
**Pull request by skrchnavy**
* `Session.attrs` table enhanced to support getter and setter methods, not only values
* `VI_ATTR_TMO_VALUE` attribute handling unified to setup also self.timeout as value holder (can be used in sessions and is set to 'python' timeout - seconds as float or None for infinite)
  * `_get_timeout`/`_set_timeout` methods added and updated for places where external lib timeout is setup.
  * GPIBInstrSessiontimeout handling corrected (tested without HW if correct value is calculated in different conditions)
* TCPIP Socket Session supports additional attributes (were not implemented)
  * `Session.attrs` table used for implementation with getters and setters
  * code cleanup to remove class attributes and use instance attributes
  * `_pending_buffer` is main buffer for handling received data, removed copying from temp buffer (`out`)  and pending buffer

### CODE REVIEW

**Comment by skrchnavy** [comment]
@MatthieuDartiailh please review.

**Comment by skrchnavy** [comment]
@MatthieuDartiailh code tested and ready for final review

**Comment by MatthieuDartiailh** [comment]
I may need some time to review this. Hopefully I will merge sometimes next week.

File: pyvisa-py/tcpip.py [hunk 1]
```diff
@@ -335,19 +329,26 @@ def after_parsing(self):
         # TODO: board_number not handled
 
         ret_status = self._connect()
-        if ret_status != constants.StatusCode.success:
+        if ret_status != SUCCESS:
```
[the remark this answers is not in the document]
Comment (skrchnavy) [reply pyvisa-py/tcpip.py]: This is unification of return codes in this file with previous implementation, when SUCCESS was used, one way shall be followed, so which of these 3 options (see line 25 and 26 of this file):
* `SUCCES`
* `StatusCode.success`
* `constants.StatusCode.success`
it is related other return codes in the file, when second form is used.

File: pyvisa-py/sessions.py [hunk 2]
```diff
@@ -309,7 +313,8 @@ def _read(self, reader, count, end_indicator_checker, suppress_end_en,
         # NOTE: Some interfaces return not only a single byte but a complete block for each read
         # therefore we must handle the case that the termination character is in the middle of the  block
         # or that the maximum number of bytes is exceeded
-        timeout = self.get_attribute(constants.VI_ATTR_TMO_VALUE)[0] / 1000.
+        timeout, _ = self.get_attribute(constants.VI_ATTR_TMO_VALUE)
```
[the remark this answers is not in the document]
Comment (skrchnavy) [reply pyvisa-py/sessions.py]: Unified form of reading timeout in all *Session object

But now I seethis shall be changed to use `self.timeout, it is now properly filled containing python way of timeout (seconds + None) so can be used directly without double conversion.

**Comment by MatthieuDartiailh** [comment]
One stupid question: shouldn't `timeout` on the TCPIP session be a private attribute rather than a public one ? From my understanding it should only be set through the visa attribute and not directly.

**Comment by skrchnavy** [comment]
Further changes / fixes in timeout handling added, some comments applied).

Regarding `Session.timeout` attribute: it is protected but also `Session.interface` or `Session.parsed` are protected.  Candidate for writing new issue for this kind of refactoring.

How to cope with the `SUCCESS` naming convention? spread across Sessions. Which is prefered? (I like `StatusCode.success` way as it is used for reast of status codes  and short enough.)

File: pyvisa-py/sessions.py [hunk 3]
```diff
@@ -181,7 +184,8 @@ def __init__(self, resource_manager_session, resource_name, parsed=None, open_ti
         self.attrs = {constants.VI_ATTR_RM_SESSION: resource_manager_session,
```
[the remark this answers is not in the document]
Comment (skrchnavy) [reply pyvisa-py/sessions.py]: Comment added into `after_parsing` method of Session

**Comment by skrchnavy** [comment]
@MatthieuDartiailh #118 have to be resolved (merged or closed) before this issue (in conflict)

File: pyvisa-py/tcpip.py [hunk 4]
```diff
@@ -401,56 +404,60 @@ def read(self, count):
...
-                term_byte_index = out.index(term_byte) + 1
-                self._pending_buffer = out[term_byte_index:]
-                return bytes(out[:term_byte_index]), constants.StatusCode.success_termination_character_read
-
-            if len(out) >= count:
-                self._pending_buffer = out[count:]
-                return bytes(out[:count]), constants.StatusCode.success_max_count_read
+            if term_char_en and term_byte in self._pending_buffer:
+                term_byte_index = self._pending_buffer.index(term_byte) + 1
+                if term_byte_index > count:
```
Comment (skrchnavy) [inline pyvisa-py/tcpip.py]: this check improves handling of case when requested less data then recevied and also terchar received. (#101)

File: pyvisa-py/gpib.py [hunk 5]
```diff
@@ -41,10 +39,9 @@ def _find_listeners():
 
 StatusCode = constants.StatusCode
 
-# linux-gpib timeout constants, in milliseconds. See self.timeout.
-TIMETABLE = (0, 1e-2, 3e-2, 1e-1, 3e-1, 1e0, 3e0, 1e1, 3e1, 1e2, 3e2, 1e3, 3e3,
-             1e4, 3e4, 1e5, 3e5, 1e6)
-
+# linux-gpib timeout constants, in milliseconds. See GPIBSession._set_timeout.
```
Comment (MatthieuDartiailh) [inline pyvisa-py/gpib.py]: Those numbers are in second now no ?

Comment (skrchnavy) [reply pyvisa-py/gpib.py]: yes, i forgot to change comment properly

**Comment by MatthieuDartiailh** [comment]
Merging. Thanks again @skrchnavy 

bors r+

**Pull request merged**